src/H5Dio.c

```diff
@@ -137,363 +131,597 @@ H5D__read(H5D_t *dataset, hid_t mem_type_id, H5S_t *mem_space, H5S_t *file_space
-            HGOTO_ERROR(H5E_PLIST, H5E_CANTGET, FAIL, "can't tell if fill value defined")
-
-        /* Should be impossible, but check anyway... */
-        if (fill_status == H5D_FILL_VALUE_UNDEFINED &&
-            (dataset->shared->dcpl_cache.fill.fill_time == H5D_FILL_TIME_ALLOC ||
-             dataset->shared->dcpl_cache.fill.fill_time == H5D_FILL_TIME_IFSET))
-            HGOTO_ERROR(H5E_DATASET, H5E_READERROR, FAIL,
-                        "read failed: dataset doesn't exist, no data can be read")
-
-        /* If we're never going to fill this dataset, just leave the junk in the user's buffer */
-        if (dataset->shared->dcpl_cache.fill.fill_time == H5D_FILL_TIME_NEVER)
-            HGOTO_DONE(SUCCEED)
-
-        /* Go fill the user's selection with the dataset's fill value */
-        if (H5D__fill(dataset->shared->dcpl_cache.fill.buf, dataset->shared->type, buf, type_info.mem_type,
-                      mem_space) < 0)
-            HGOTO_ERROR(H5E_DATASET, H5E_READERROR, FAIL, "filling buf failed")
-        else
-            HGOTO_DONE(SUCCEED)
-    } /* end if */
+        /* If space hasn't been allocated and not using external storage,
+         * return fill value to buffer if fill time is upon allocation, or
+         * do nothing if fill time is never.  If the dataset is compact and
+         * fill time is NEVER, there is no way to tell whether part of data
+         * has been overwritten.  So just proceed in reading.
+         */
+        if (dset_info[i].nelmts > 0 && dset_info[i].dset->shared->dcpl_cache.efl.nused == 0 &&
+            !(*dset_info[i].dset->shared->layout.ops->is_space_alloc)(
+                &dset_info[i].dset->shared->layout.storage) &&
+            !(dset_info[i].dset->shared->layout.ops->is_data_cached &&
+              (*dset_info[i].dset->shared->layout.ops->is_data_cached)(dset_info[i].dset->shared))) {
+            H5D_fill_value_t fill_status; /* Whether/How the fill value is defined */
+
+            /* Retrieve dataset's fill-value properties */
+            if (H5P_is_fill_value_defined(&dset_info[i].dset->shared->dcpl_cache.fill, &fill_status) < 0)
+                HGOTO_ERROR(H5E_PLIST, H5E_CANTGET, FAIL, "can't tell if fill value defined")
+
+            /* Should be impossible, but check anyway... */
+            if (fill_status == H5D_FILL_VALUE_UNDEFINED &&
+                (dset_info[i].dset->shared->dcpl_cache.fill.fill_time == H5D_FILL_TIME_ALLOC ||
+                 dset_info[i].dset->shared->dcpl_cache.fill.fill_time == H5D_FILL_TIME_IFSET))
+                HGOTO_ERROR(H5E_DATASET, H5E_READERROR, FAIL,
+                            "read failed: dataset doesn't exist, no data can be read")
+
+            /* If we're never going to fill this dataset, just leave the junk in the user's buffer */
+            if (dset_info[i].dset->shared->dcpl_cache.fill.fill_time == H5D_FILL_TIME_NEVER)
+                HGOTO_DONE(SUCCEED)
```

Comment:
Do we want this to be a goto_done or a continue? As in, do we expect that different datasets in the I/O operation could have different fill times? I'm assuming here that we don't want to short-circuit the whole I/O in the case that a single dataset has a fill time of never

Reply:
Fixed, thanks. Was more complicated but I think I've gotten it right. Needs a test case, I'll add one soon.